source/iterator/snapshot.go

```diff
@@ -0,0 +1,198 @@
+func (s *Snapshot) HasNext(ctx context.Context) (bool, error) {
+	if s.rows != nil && s.rows.Next() {
+		return true, nil
+	}
+
+	if err := s.loadRows(ctx); err != nil {
+		return false, fmt.Errorf("load rows: %w", err)
+	}
+
+	return false, nil
+}
+
+// Next returns the next record.
+func (s *Snapshot) Next(ctx context.Context) (sdk.Record, error) {
+	row := make(map[string]any)
+	if err := s.rows.MapScan(row); err != nil {
+		return sdk.Record{}, fmt.Errorf("scan rows: %w", err)
+	}
+
+	transformedRow, err := coltypes.TransformRow(row, s.columnTypes)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("transform row column types: %w", err)
+	}
+
+	if _, ok := transformedRow[s.orderingColumn]; !ok {
+		return sdk.Record{}, errOrderingColumnIsNotExist
+	}
+
+	pos := Position{
+		Mode:             ModeSnapshot,
+		LastProcessedVal: transformedRow[s.orderingColumn],
+		//Time:             time.Now(),
+	}
+
+	convertedPosition, err := pos.convertToSDKPosition()
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("convert position %w", err)
+	}
+
+	if _, ok := transformedRow[s.keyColumn]; !ok {
+		return sdk.Record{}, errKeyIsNotExist
+	}
+
+	transformedRowBytes, err := json.Marshal(transformedRow)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("marshal row: %w", err)
+	}
+
+	s.position = &pos
+
+	return sdk.Record{
+		Position: convertedPosition,
+		Metadata: map[string]string{
+			metadataTable:  s.table,
+			metadataAction: string(actionInsert),
+		},
+		CreatedAt: time.Now(),
+		Key: sdk.StructuredData{
+			s.keyColumn: transformedRow[s.keyColumn],
+		},
+		Payload: sdk.RawData(transformedRowBytes),
+	}, nil
+}
+
+// Ack check if record with position was recorded.
+func (s *Snapshot) Ack(ctx context.Context, position sdk.Position) error {
+	sdk.Logger(ctx).Debug().Str("position", string(position)).Msg("got ack")
+
+	return nil
+}
+
+// Stop stops snapshot iterator.
+func (s *Snapshot) Stop() error {
+	if s.rows != nil {
+		err := s.rows.Close()
+		if err != nil {
+			return err
+		}
+	}
+
+	if s.db != nil {
+		return s.db.Close()
+	}
+
+	return nil
+}
+
+// LoadRows selects a batch of rows from a database, based on the Iterator's
+// table, columns, orderingColumn, batchSize and the current position.
+func (s *Snapshot) loadRows(ctx context.Context) error {
+	selectBuilder := sqlbuilder.NewSelectBuilder().
+		From(s.table).
+		OrderBy(s.orderingColumn)
+
+	if len(s.columns) > 0 {
+		selectBuilder.Select(s.columns...)
+	} else {
+		selectBuilder.Select("*")
+	}
+
+	if s.position != nil {
+		selectBuilder.Where(
+			selectBuilder.GreaterThan(s.orderingColumn, s.position.LastProcessedVal),
+		)
+	}
+
+	query, err := sqlbuilder.DefaultFlavor.Interpolate(
+		sqlbuilder.Buildf("%v FETCH NEXT %v ROWS ONLY", selectBuilder, s.batchSize).Build(),
+	)
+	if err != nil {
+		return fmt.Errorf("interpolate arguments to SQL: %w", err)
+	}
+
+	rows, err := s.db.QueryxContext(ctx, query)
```

Comment:
Could we do it with args? rows, err := s.db.QueryxContext(ctx, query, args ...).
I asked it because it this case can be problem with quotes values. if it is fine with quotes it is not needed

Reply:
I've tested this with the next values:
1. double quote `test " test \"`
2. single quote `test ' test \'`
And it works.

Reply:
What about oracle system words?

Reply:
Do you mean keywords kind of `insert` as a field's value?
It works also.